### DISCUSSION TITLE
Sunshine installer is very rude and abruptly reboots your PC without any confirmation if ViGEmBus is already installed

### DISCUSSION THREAD

**Discussion by Dwedit** [LizardByte/community#319] [Sunshine] [answered]
I had ViGEmBus installed on the PC already from another program (DS4windows), but the installer for Sunshine abruptly rebooted my PC (no confirmation dialog) when it reached the part of the installer when it installs ViGEmBus.  This is not acceptable.

**Comment by ReenigneArcher** [comment]
When Sunshine installs Vigembus, there may be a prompt for a restart if required.

https://github.com/LizardByte/Sunshine/blob/7166d2f1fd9fbd35a687a9694cf30c0e9e1c18ef/src_assets/windows/misc/gamepad/install-gamepad.bat#L46

There is nothing else anywhere in the installer that would force a restart... and no restart is required for anything other than Vigembus.

I have personally installed Sunshine hundreds of times (with vigembus), and never had my PC restarted by the installation.

**Reply by ReenigneArcher** [reply]
@nefarius Do you have some idea about this?

We're installing "ViGEmBus" as part of the Sunshine install, with the command above. I've never had it happen to me in all my test installs, but a few users are reporting their computer is being forcefully restarted, with no prompt... even though I have specified to prompt for a restart.

**Reply by nefarius** [reply]
Which version do you bundle? Is that my unaltered setup? If so `/promptrestart` is not a recognised CLI arg, IDK where you got that from.

Also out of curiosity; where are you crediting ViGEmBus' involvement, couldn't find it anywhere...

**Reply by ReenigneArcher** [reply]
The latest version that's available on GitHub releases.

I got it from the help arg of the installer, although this was some time ago.

There's a reference here: https://docs.lizardbyte.dev/projects/sunshine/en/latest/about/usage.html#windows

There used to be far more references, but due to the trademark issues we renamed everything to "virtual gamepad" for now.

**Reply by nefarius** [reply]
> There's a reference here: https://docs.lizardbyte.dev/projects/sunshine/en/latest/about/usage.html#windows
> 
> There used to be far more references, but due to the trademark issues we renamed everything to "virtual gamepad" for now.

Gotcha, thanks. Yeah the situation for the future of users of the framework is completely unknown to me as well. Time will tell.

[Here are the supported command line arguments](https://www.advancedinstaller.com/silent-install-exe-msi-applications.html), I checked the latest setup code, there is nothing in it that could force a reboot **unless** the user had an older version with a devcon bug installed and the upgrade triggered the uninstaller of the old version, which indeed had an unintentional force reboot in it.

**TL;DR:** Sunshine installer is actually *not* rude 🤣 you simply hit a bug from a long time ago. Always keep your software up2date, folks 🖖

**Reply by ReenigneArcher** [reply]
Thanks for the info!

The below is just for future reference.

Options for the exe installer.
![image](https://github.com/LizardByte/Sunshine/assets/42013603/54adb0ce-4d81-413a-ada9-cfd5a1891e25)

The final line mentions it accepts msiexec.exe arguments as well, which are:

```txt
Windows ® Installer. V 5.0.22621.1 

msiexec /Option <Required Parameter> [Optional Parameter]

Install Options
	</package | /i> <Product.msi>
		Installs or configures a product
	/a <Product.msi>
		Administrative install - Installs a product on the network
	/j<u|m> <Product.msi> [/t <Transform List>] [/g <Language ID>]
		Advertises a product - m to all users, u to current user
	</uninstall | /x> <Product.msi | ProductCode>
		Uninstalls the product
Display Options
	/quiet
		Quiet mode, no user interaction
	/passive
		Unattended mode - progress bar only
	/q[n|b|r|f]
		Sets user interface level
		n - No UI
		b - Basic UI
		r - Reduced UI
		f - Full UI (default)
	/help
		Help information
Restart Options
	/norestart
		Do not restart after the installation is complete
	/promptrestart
		Prompts the user for restart if necessary
	/forcerestart
		Always restart the computer after installation
Logging Options
	/l[i|w|e|a|r|u|c|m|o|p|v|x|+|!|*] <LogFile>
		i - Status messages
		w - Nonfatal warnings
		e - All error messages
		a - Start up of actions
		r - Action-specific records
		u - User requests
		c - Initial UI parameters
		m - Out-of-memory or fatal exit information
		o - Out-of-disk-space messages
		p - Terminal properties
		v - Verbose output
		x - Extra debugging information
		+ - Append to existing log file
		! - Flush each line to the log
		* - Log all information, except for v and x options
	/log <LogFile>
		Equivalent of /l* <LogFile>
Update Options
	/update <Update1.msp>[;Update2.msp]
		Applies update(s)
	/uninstall <PatchCodeGuid>[;Update2.msp] /package <Product.msi | ProductCode>
		Remove update(s) for a product
Repair Options
	/f[p|e|c|m|s|o|d|a|u|v] <Product.msi | ProductCode>
		Repairs a product
		p - only if file is missing
		o - if file is missing or an older version is installed (default)
		e - if file is missing or an equal or older version is installed
		d - if file is missing or a different version is installed
		c - if file is missing or checksum does not match the calculated value
		a - forces all files to be reinstalled
		u - all required user-specific registry entries (default)
		m - all required computer-specific registry entries (default)
		s - all existing shortcuts (default)
		v - runs from source and recaches local package
Setting Public Properties
	[PROPERTY=PropertyValue]

Consult the Windows ® Installer SDK for additional documentation on the
command line syntax.

Copyright © Microsoft Corporation. All rights reserved.
Portions of this software are based in part on the work of the Independent JPEG Group.
```

**Reply by nefarius** [reply]
`/exenoui /quiet /norestart` sounds sufficient.

**Reply by ReenigneArcher** [reply]
Yea, but Sunshine doesn't recognize game pads unless the computer is restarted after the first install of Vigembus.

**Reply by nefarius** [reply]
Weird.

**Comment by Dwedit** [comment]
Well I didn't get a prompt when I just installed it, just Windows 10 suddenly going into reboot mode with no prompt at all.

Sorry if I'm being too rude or angry here.  My computer just got rebooted and I'm a little cranky.

**Reply by ReenigneArcher** [reply]
Something else restarted your computer. Sunshine didn't.

**Reply by cgutman** [reply, marked as answer]
I think it's the ViGEmBus installer doing it. It happened to me months ago, but I could never reproduce it. I think it may happen when ViGEmBus is in use by another application and we try to update it.

**Comment by Coolykoen** [comment]
It is certainly the vigembus installer. I just had it happen, when updating sunshine, and i too already had vigembus, the msi installer for vigembus popped up and about 2 seconds later it restarted the computer (not crash, just "restarting...").
I would suggest making the vigembus installer optional. "vigembus is required for gamepad support, do you want to install it? It may restart your computer so save any open documents."

Just force rebooting is indeed not acceptable as the topic starter mentioned.

**Reply by ReenigneArcher** [reply]
There's no intention of force rebooting... as I previously indicated we specify a prompt for restart.

https://github.com/LizardByte/Sunshine/discussions/1718#discussioncomment-7220314

Personally, in my hundreds of installs... I've never had it restart or prompt.

> I would suggest making the vigembus installer optional.

It is optional. Just uncheck "virtual gamepad".

**Reply by Mansarde** [reply]
Just had this happen to me as well. I had so much stuff open, all mercilessly stomped into the ground by this.
Pain. 😫

Thanks for the hint with the virtual gamepad <-> vigembus.

**Reply by ReenigneArcher** [reply]
Unfortunately this bug returned in the latest/final release of vigembus. 😩

The only thing we can do really is go back to not installing it, and answer 20 million questions a day about why gamepad input isn't working.

**Reply by Dwedit** [reply]
Still one other thing you can do...

Run the installer as a suspended process, detour any API calls that shutdown the computer (such as ExitWindowsEx), then resume the process.

But I'm still not sure exactly which process is doing the system reboot.  It could be the installer, or it could be something installed by the installer.
